src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java

```diff
@@ -109,46 +126,75 @@ public void produceHourlySnapshotsForOrg(String orgId, DateRange snapshotRange)
     // fetched by service type just once.
     Set<String> serviceTypes = SubscriptionDefinition.getAllServiceTypes();
     for (String serviceType : serviceTypes) {
-      log.info(
-          "Producing hourly snapshots for orgId {} for service type {} "
-              + "between startDateTime {} and endDateTime {}",
-          orgId,
-          serviceType,
-          snapshotRange.getStartString(),
-          snapshotRange.getEndString());
+      log.info("Producing hourly snapshots for orgId {} for service type {} ", orgId, serviceType);
+
       try {
-        var result =
-            retryTemplate.execute(
-                context -> metricUsageCollector.collect(serviceType, orgId, snapshotRange));
-        if (result == null) {
-          continue;
+        TallyState currentState =
+            tallyStateRepository
+                .findById(new TallyStateKey(orgId, serviceType))
+                .orElseGet(
+                    () ->
+                        tallyStateRepository.save(
+                            new TallyState(orgId, serviceType, clock.startOfToday().minusDays(1))));
+
+        AccountUsageCalculationCache calcCache = new AccountUsageCalculationCache();
+        while (true) {
```

Comment:
I'm a bit scared every time I see `while (true)` statements :)
What about counting the number of events to process and using a `for` statement with this count?

Reply:
I agree. I'll come up with something a little better.

Reply:
Please see if the latest changes are a little better.

Reply:
I tried a couple variations of using a `for` loop but it got  a little messy. Hopefully this is a little better.

Reply:
My main concern is that for some reason the events from `fetchEventsInBatch` are never empty, so the loop is going infinite. 
I'm unsure if a safe-ward would make sense here (a limit within the loop). I will propose another solution later today.

Reply:
@mstead this is what I had in mind: https://privatebin.corp.redhat.com/?078a45c59b426048#EizuYWrzdKfzYtnxEGGytCKT31kPMzp7kU8tJV2DoCbf

Basically, stream the events (using QueryHints) instead of querying the database every X times until we have no more data). I've done this via the stream support of Spring JPA.

In the patch from the pastebin link, I'm doing all the changes in a single transaction. This needs to be refactored in separate transactions as you stated in this comment https://github.com/RedHatInsights/rhsm-subscriptions/pull/2876/files#diff-8005db7bcd7bb68e8259f0d4a7f02e0a0797bf75bb380b8c14cb47e5bf22ce75R114.

Reply:
Ya, I hear you for sure, and honestly, this is something that I was thinking about while doing this. Thinking about the processing 1 million Events benchmark, batching allows us to also lookup the hosts related to the Event in the batch all at once, which is why I leaned this way.

> A disadvantage of using streams is that we can't customize the count of the batches using the hourlyTallyEventBatchSize property, but it's hardcoded to the value set in the QueryHint annotation. I've checked that we're using QueryHint in other streaming operations, so I think we can live with that.
Agreed.

> List<Event> events = eventRecords.map(EventRecord::getEvent).toList();

Unless I'm mistaken, the above line would pull all of the EventRecords into memory all at once. We could process the events one by one, but I'm not sure what it would do to the performance of Host updates.

Thanks for the suggestions! Gives me some things to investigate. I'll look into this some more today.

Reply:
> > List events = eventRecords.map(EventRecord::getEvent).toList();
> 
> Unless I'm mistaken, the above line would pull all of the EventRecords into memory all at once. We could process the events one by one, but I'm not sure what it would do to the performance of Host updates.

Not all the events, but only the ones that are part of the current batch of the stream. 
Semantically, this is the same as you are doing now, but avoiding having to query several times to check whether there are more events to process.

Reply:
@Sgitario I pushed an update that I think addresses your concern as a separate commit to make it easier to follow. It continues to support batch processing the events, however, it now streams from the DB. I tested the effect of changing the batch size of the Stream and it had less of an impact of adjusting the processing batch size, so I think it is fine as is, and still provides the flexibility of adjusting the overall processing batch size. It also keeps the transaction setup the same.

Let me know what you think.